### DISCUSSION TITLE
How can I dispose of these cache instances entirely?

### DISCUSSION THREAD

**Discussion by trung** [ben-manes/caffeine#1790] [Q&A] [answered]
I create temporary caches dynamically for each request. When a request completes, the cache items are evicted, leaving the cache instances empty but still in memory.

How can I dispose of these cache instances entirely?

**Comment by ben-manes** [comment, marked as answer]
I'm not sure what you mean. The cache does not maintain any global state, e.g. we do not directly offer a `CacheManager`. Caffeine is just like a `java.util.ConcurrentHashMap` except with a bounding policy, so once there are no more references then the GC will collect. If you create an instance per request then when the request completes, the cache should go away along with any other request-scoped state.

If you use a higher level abstraction, like a CacheManager, which is assuming singletons that are long-lived and reused, then you would need to unregister the instance. That might be more of a Spring, JCache, etc concept which has expectations of how you use their frameworks. We are not so opinionated, Caffeine is merely a data structure library and not an application framework.

Can you explain more about what you are doing? You might need to ask on a different forum like the application framework's as more their world than ours.

**Reply by trung** [reply]
Thanks, I'm using `CaffeineCacheManager` abstraction from Spring and it maintains a `cacheMap` which keeps growing as caches are dynamically created. I'll go over to Spring forums to see how a cache can be deregistered.

**Reply by ben-manes** [reply]
my best guess is that you would mark the `CacheManager` as `@RequestScoped` so that, while the instance is not deregistered, the entire manager goes away. but alas I have not actually used Spring since 1.x in its xml heydays, prior to any caching support.

**Reply by trung** [reply]
https://github.com/spring-projects/spring-framework/issues/33813
